metric_learn/constraints.py

```diff
@@ -33,6 +35,110 @@ def positive_negative_pairs(self, num_constraints, same_length=False,
+    same class (X_b) and taking one of its `k_impostor`-nearest neighbors of
+    other classes (X_c).
+
+    In the case a class doesn't have enough points in the same class (other
+    classes) to yield `k_genuine` (`k_impostor`) neighbors a warning will be
+    raised and the maximum value of genuine (impostor) neighbors will be used
+    for that class.
+
+    Parameters
+    ----------
+      X : (n x d) matrix
+        Input data, where each row corresponds to a single instance.
+      k_genuine : int
+        Number of neighbors of the same class to be taken into account.
+      k_impostor : int
+        Number of neighbors of different classes to be taken into account.
+
+    Returns
+    -------
+    triplets : array-like, shape=(n_constraints, 3)
+      2D array of triplets of indicators.
+    """
+
+    labels, labels_count = np.unique(self.partial_labels, return_counts=True)
+    n_labels = len(labels)
+    len_input = np.size(self.partial_labels, 0)
+
+    # Handle the case where there are too few elements to yield k_genuine or
+    # k_impostor neighbors for every class.
+
+    k_genuine_vec = np.ones(n_labels, dtype=np.intp)*k_genuine
+    k_impostor_vec = np.ones(n_labels, dtype=np.intp)*k_impostor
+
+    for i in range(n_labels):
+      if (k_genuine + 1 > labels_count[i]):
+        k_genuine_vec[i] = labels_count[i]-1
+        warnings.warn("The class {} has {} elements, which is not sufficient "
+                      "to generate {} genuine neighbors as specified by "
+                      "k_genuine. Will generate {} genuine neighbors instead."
+                      "\n"
+                      .format(labels[i], labels_count[i], k_genuine+1,
+                              k_genuine_vec[i]))
+      if (k_impostor > len_input - labels_count[i]):
+        k_impostor_vec[i] = len_input - labels_count[i]
+        warnings.warn("The class {} has {} elements of other classes, which is"
+                      " not sufficient to generate {} impostor neighbors as "
+                      "specified by k_impostor. Will generate {} impostor "
+                      "neighbors instead.\n"
+                      .format(labels[i], k_impostor_vec[i], k_impostor,
+                              k_impostor_vec[i]))
+
+    triplets = np.empty((np.dot(k_genuine_vec*k_impostor_vec, labels_count),
+                         3), dtype=np.intp)
+
+    start = 0
+    finish = 0
+    neigh = NearestNeighbors()
+
+    for i in range(n_labels):
+
+        # generate mask for current label
+        gen_mask = self.partial_labels == labels[i]
+        gen_indx = np.where(gen_mask)
+
+        # get k_genuine genuine neighbours
+        neigh.fit(X=X[gen_indx])
+        gen_neigh = np.take(gen_indx, neigh.kneighbors(
+                            n_neighbors=k_genuine_vec[i],
+                            return_distance=False))
+
+        # generate mask for impostors of current label
+        imp_indx = np.where(np.invert(gen_mask))
```

Comment:
`imp_mask = ~gen_mask`

Reply:
You mean to separate it in two lines for clarity or just replacing the `np.invert` by the `~`?

Reply:
I think you can avoid creating `imp_indx` entirely, as it's only used for indexing as far as I can see.

Reply:
It is used later to create `imp_neigh` on line 131:
```
imp_neigh = np.take(imp_indx, neigh.kneighbors(n_neighbors=k_impostor_vec[i], X=X[gen_mask], return_distance=False))
```
This maps the indeces of the impostors neighbors yielded by `niegh.kneighbors() ` that reference `X[imp_indx]` back to the indeces of X.

Maybe this is a bit unclear, should I refactor it to be more understandable?

Reply:
Ooh, I misread that line at first. I do think it would be clearer if written as a plain indexing expression:

```
neighbor_inds = imp_indx[neigh.kneighbors(n_neighbors=k_impostor_vec[i], X=X[gen_mask], return_distance=False)]
```

Reply:
Sadly it cannot be indexed like that! as the `neigh.kneighbors()` yields a vector of indices of n_input times k. I changed it a little bit for clarity and added a comment, let me know what you think!